WIP: add vc-got-annotate

zmyrgel opened the pull request:
Heads up, I'm working on adding annotate functionality.
This initial version adds the few annotate commands. I'm working on getting next/prev-revision added to this so annotate buffer can show diffs etc.

zmyrgel committed:
add vc-got-annotate-{command,time,extract-revision-at-line}

this makes vc-annotate command basic functionality working, needs
next/prev-revision support to see diffs.

zmyrgel committed:
add vc-got-{previous,next}-revision defuns

roughy working but might need a bit more testing.

zmyrgel committed:
add vc-got-annotate-re helper

omar-polo commented on vc-got.el:
```diff
@@ -593,5 +593,33 @@ Value is returned as floating point fractional number of days."
+  (let (process-file-side-effects)
+    (vc-got-with-worktree file
+      (with-temp-buffer
+        (when (zerop (apply #'vc-got--call "log" (list "-c" rev "-l" "2" "-R" file)))
```
I think you can use `vc-got--log` instead of applying `vc-got--call` and building a list by hand.  Sure, the code of `vc-got--log` can probably be simplified, but that's another issue :)

The advantage is also that vc-got--log will implicitly change the worktree, so you can also avoid the `vc-got-with-worktree`, and returns t or nil, so you can avoid also the zerop.  As an additional improvement we can move the `(let (process-file-side-effects) ...)` to `vc-got--log` to further reduce the boilerplate (from a quick check the log subcommand cannot alter the content of the source files in the worktree),

omar-polo replied:
I forgot to mention it in the commit message of 0a66694e897e35f0472d5ada455a76b7719cc285, but I added an argument for `vc-got--log` to require the log messages in reverse order (the "-R" flag) and to bind `process-file-sed-effects` :)

zmyrgel replied:
I was trying following in vc-got-previous-revision:

```
(defun vc-got-previous-revision (file rev)
  "Return the revision number that precedes REV for FILE, or nil if no such revision exists."
  (with-temp-buffer
    (vc-got--log file 2 rev nil t)
    (keep-lines "^commit")
    (when (looking-at "^commit \\([a-z0-9]+\\)")
      (match-string-no-properties 1))))
```

But that gives error "Wrong type argument: stringp, t" and seems to point to rev argument to log for some reason.

omar-polo replied:
you forgot a nil before the last t.  the arguments now are `(path limit start-commit stop-commit search-patter reverse)`, so `(vc-got--log file 2 rev nil nil t)` should work.

omar-polo replied:
btw, the arguments for `vc-got--log` are perhaps too much, maybe it's time to switch to keyword arguments.

omar-polo reviewed (commented):
Thanks for working on this.  It's one of the most interesting features of VC in my opinion.

The patch looks good to me, I quickly tried the vc-annotate function against a couple of files and it seems to work.  I couldn't use the previous/next-revision functions, but I'll give another shot soon.

I'll push a commit soon to change `vc-got--log` so it accepts an additional argument to reverse the commit list, so you can switch `vc-got-previous-revision` to it.

Cheers!

zmyrgel committed:
use vc-got--log in prev/next-revision functions

no need to reinvent the wheel, use the existing vc-got--log to get the
log info.

zmyrgel commented:
fixed the commands to use vc-got--log. Now the next/prev can be called from annotate buffer but they won't still actually work. They just report "can't increment/decrement" from revision. I haven't still figured out why they can't do it. 
My guess is some got log call doesn't resolve the worktree and fails to give proper output leaving 'nil' return value.

omar-polo commented on vc-got.el:
```diff
@@ -605,31 +605,27 @@ Value is returned as floating point fractional number of days."
-            (match-string-no-properties 1)))))))
+  (with-temp-buffer
+    (vc-got--log file 2 rev nil nil t)
+    (keep-lines "^commit")
```
keep-lines doesn't do anything because the point is at the end of the buffer after vc-got--log.  (process-file leaves the point at the end of the output).  looking-at fails and so nil is returned.

keep-lines changes only the line past the point, so you either `(goto-char (point-min))` before keep-lines or call keep-lines with a rstart and rend argument.

Even with this I get an "Buffer \*Annotate vc-got.el (rev xyz)\* is not associated with a file" though.

zmyrgel replied:
Ah, I tested with empty buffer and run C-u M-! to run got log command and that left the point at the start of buffer. I had the goto-char before so I'll restore that there.

About the error I think the vc-got-annotate-extract-revision-at-line should return (cons revision . filename) instead of just revision. That at least the what the vc-annotate-extract-revision-at-line doc string says but there's different comment on vc.el. There its said "If the backend supports annotating through copies and renames, and displays a file name and a revision, then return a cons (REVISION . FILENAME)." I don't know what it tries to say exactly but I think its worth to try to add the filename as its return value.

omar-polo replied:
I'm not sure it's related to the error, but yeah, annotate-extract-revision-at-line should return `(revision . filename)`.  Unlike CVS, git (and thus got) supports renaming files and probably helps when the previous/next revision points to a file with a different name.  (to be fair, from a quick look only vc-git seems to return a cons, vc-svn and vc-hg don't)

I'm not sure how to get the filename though.  One idea could be define a buffer-local variable and set it in vc-got-annotate-command.  Something along the lines of

```elisp
(defvar-local vc-got--annotate-file nil
  "The file being annotated")

(defun vc-got-annotate-command (file buf &optional rev)
  "..."
  (with-current-buffer buf
    (setq vc-got--annotate-file file)
    ...))

(defun vc-got-annotate-extract-revision-at-line ()
  ...
  (cons (match-string-no-properties 1) vc-got--annotate-file))
```

but it seems pretty hackish.  (Also, FYI `with-current-buffer` is equivalent to `(save-current-buffer (set-buffer buf) ...)`)

Maybe we can ask the devs of got to include the filename in the output of blame?  Not sure if it's a good approach though.

zmyrgel replied:
It would seem a bit redundant to print filename in blame, you give the path as argument to blame so the caller should know it.

I don't think setting local variable would work, what if user lists annotation for another file between toggleing prev/next in old file. It would list wrong files contents.
How about this, parse the filename from the buffer name?

```
(defun vc-got-annotate-extract-revision-at-line ()
  "Returns revision corresponding to the current line or nil."
  (save-excursion
    (beginning-of-line)
    (when (string-match "*Annotate \\(.+\\) (rev [a-z0-9]+)\\*$" (buffer-name))
      (let ((fname (match-string-no-properties 1)))
        (when (looking-at vc-got-annotate-re)
          (cons (match-string-no-properties 1)
                (expand-file-name fname (vc-got-root default-directory))))))))
```

omar-polo replied:
There's a problem with parsing the buffer name.  The buffer name contains only the filename, not the path.  Just tried on a project: visited `src/clj/foo/search.clj` and the annotate buffer (from vc-git) is called "\*Annotate search.clj\*".  If I have another `search.clj` in that project, say `test/foo/search.clj` you don't know which of those two is the one I'm interested in.

An aside, sometimes I got confused by the buffer-local variables, but I don't think that your example is right.  if I run vc-annotate on file1 I get a buffer "Annotate file1".  Then I run vc-annotate on file2, and I get a buffer "Annotate file2".   Now, vc-got--annotate-file is "file1" in "Annotate file1" and "file2" in buffer "Annotate file2", since it's buffer local.  it's bound to that buffer, regardless if I change it on another IIUC.

Anyway, all of this is meaningless.  I just remembered reading about a `(vc-deduce-fileset)` while reading some diff-related code.  It seems that (at least for vc-git, I don't have the time to further test right now) it yields the full path to the file if called in an annotation buffer.  Try it with M-: (vc-deduce-fileset) on a annotation buffer.

omar-polo replied:
nevermind, `(vc-deduce-fileset)` doesn't seem to work.  But at least we can access the original file through the buffer-local `vc-annotate-parent-file`: vc does the hack for us!  (now the question is: given that c-annotate-parent-file` is provided by VC, why it requires us to return `(rev . file)` when it already knows the filename? )

Anyway, I'm curios why we have cases where VC errors about an annotated buffer that isn't associated with a file.

omar-polo replied:
(sorry for the spam)

Eureka!  I think I got why `annotate-extract-revision-at-line` may return `(rev . file)`.  The problem is not getting the original file name, in fact vc does that by itself (in `vc-annotate` it set a local variable for that), but to handle the case when file gets renamed.  It makes sense for got (or any other vcs really) to output the filename in the blame output because a line in `file1` may come from `file2`, if at some point in the history `file2` was renamed to `file1`.

Anyway, this shouldn't block this PR.  For the time being we can simply return the revision, as you're doing.  If you don't beat me to it, between this night and tomorrow morning I should find the time to debug the remaining issues in the next/prev revision functions.

zmyrgel replied:
Got previous to semi-working by adding (setq buffer-file-name file) to vc-got-annotate-command and then returning it from annotate-extract-revision-at-line. Though it messed the buffer name for some reason. It started to loop "Annotate Annotate Annotate .init.el ...".

omar-polo replied:
Yeah, I spent some time debugging this and part of the fault is due to `vc-ensure-vc-buffer` not recognising the annotate buffer as managed by got.  This is probably my fault for not using `vc-do-command`, I still have to investigate that.

I managed to get next and previous revision working, I'll add comments to indicate the modification I had to do.

omar-polo commented on vc-got.el:
```diff
@@ -605,31 +605,27 @@ Value is returned as floating point fractional number of days."
+    (vc-got--log file nil nil rev)
+    (keep-lines "^commit")
+    (goto-char (point-max))
+    (beginning-of-line)
+    (previous-line) ;; return from empty line to last actual commit
```
instead of `(beginning-of-line)` plus `(previous-line)` you can do `(forward-line -1)`.  Also, but don't ask me why, `previous-line` should be used only interactively (or at least this is the warning I'm getting)

zmyrgel replied:
Ok, I haven't read the elisp info pages yet so I'm still not familiar with the elisp idioms.

omar-polo replied:
No worries :)

omar-polo commented on vc-got.el:
```diff
@@ -568,5 +568,64 @@ LIMIT limits the number of commits, optionally starting at START-REVISION."
+    (goto-char (point-max))
+    (beginning-of-line)
+    (previous-line) ;; return from empty line to last actual commit
+    (when (looking-at  "^commit \\([a-z0-9]+\\)$")
```
(sorry for the spam)

the `$` shouldn't be there.  I was debugging this and realised that the point is looking at

```
commit 8da4c01ab1701f49af2bb11b89698b2b3f839f1b (master, main, github/master, origin/master)
```

zmyrgel replied:
Yeah, it was there originally to omit that commit as I think that points to to top level commit. But I think the "string= rev ..." comparison checks that so the $ is not needed.

zmyrgel committed:
fix vc-got-previous-revision

vc-got--log leaves point at the end of buffer so move point to the
beginning before running keep-lines so it has chance at working.

zmyrgel committed:
fixes for vc-got-next-revision

set rstart/end to keep-lines so it works on whole buffer and don't use
previous-line as its not intended for elisp code. Also fix the first
match regexp to allow branch names after commit sha-1 values.

zmyrgel committed:
prefer with-current-buffer

omar-polo commented on vc-got.el:
```diff
@@ -568,5 +568,64 @@ LIMIT limits the number of commits, optionally starting at START-REVISION."
+(defun vc-got-annotate-command (file buf &optional rev)
+  "Show annotated contents of FILE in buffer BUF. If given, use revision REV."
+  (let (process-file-side-effects)
+    (with-current-buffer buf
```
For the time being, we can make VC happy by

```elisp
;; FIXME: vc-ensure-vc-buffer won't recognise this buffer as managed
;; by got unless vc-parent-buffer points to a buffer managed by got.
;; investigate why this is needed.
(set (make-local-variable 'vc-parent-buffer) (find-file-noselect file))
```

and with this `vc-got-previous-revision` should work.

omar-polo commented on vc-got.el:
```diff
@@ -568,5 +568,64 @@ LIMIT limits the number of commits, optionally starting at START-REVISION."
+    (keep-lines "^commit" (point-min) (point-max))
+    (goto-char (point-max))
+    (forward-line -1) ;; return from empty line to last actual commit
+    (when (looking-at  "^commit \\([a-z0-9]+\\)")
+      ;; no need to continue if looking at top commit
+      (unless (string= rev (match-string-no-properties 1))
+        (forward-line -1)
+        (when (looking-at  "^commit \\([a-z0-9]+\\)")
+          (match-string-no-properties 1))))))
```
The first `looking-at` is redundant, because we're guaranteed to be in a line that starts with "commit" followed by a SHA1.  Actually, we're also sure that that specific SHA1 is exactly equal to the `rev` parameter, because we're asking `got log` to stop at commit with sha1 `rev`.  So the unless check should always be true.

We need to retrieve the commit id in the previous line (if it exists).  We can do that replacing the whole `when` with something like

```elisp
(unless (= (point) (point-min))
  (forward-line -1)
  (when (looking-at "^commit \\([a-z0-9]+\\)")
    (match-string-no-properties 1)))
```

forward-line will place the cursor at the beginning of the line, so we know that there is a previous line by checking that we're not at `(point-min)`.  If that's the case, it goes up one line and extract the commit id.  To be fair, here the `when` isn't needed: we called `keep-lines` before, so every line starts with "commit", but it reads nice IMHO.  You can drop the when if you prefer.

With this, I believe all the annotate commands are done?  Looking forward to merge this!

By the way, I've never used the vc-annotate command before today, but it's amazing to press `n` and `p` in that buffer and navigate in the history of the changes!

omar-polo replied:
Ops, I forgot one thing.  Can you please add a `(require 'vc-annotate)`?  It fixes an error of the byte compiler about an unknown `vc-annotate-convert-time`.  Thanks!

zmyrgel committed:
require vc-annotate

zmyrgel committed:
rename vc-got-annotate-re to vc-got--annotate-re

this is for internal use so name it so. While here add doc string for it.

zmyrgel committed:
Make annotation buffers parent managed by got

vc-ensure-vc-buffer won't recognise this buffer as managed by got unless
vc-parent-buffer points to a buffer managed by got.

zmyrgel committed:
refactor vc-got-next-revision

no need for extensive checks in function, got command won't return
successfully unless it finds the wanted revision so are certain the rev
is found, we only need to check if it has a parent.

while here, replace the multiple commit regexp with internal
vc-got--commit-re.

omar-polo commented:
Thanks!
